### ISSUE TITLE
Native Robot I/O

### ISSUE THREAD

**Issue by shaun-edwards** [ros-industrial/industrial_core#27]
This issue has been moved from: ros-industrial/ros_industrial_issues#2 (submitted by @clay-flannigan)

Native robot I/O should be supported through standard ROS-I interfaces. Most industrial controllers have built-in (embedded) I/O standard functions like fence-status, gripper-close, etc. They generally also support general purpose I/O. Exposing these I/O points would remove the requirement for separate networked I/O from the ROS PC.

**Comment by liborw** [comment]
We have already implemented an IO interface for motoman (DX100) robot. With bit of help we can make it more general.

The current implementation uses another simple message. The message is composed of pairs, IO address and command. Command represents reading, writing and read value (in case of reply from the controller).

**Comment by gavanderhoorn** [comment]
@liborw: is this code publicly accessible? I'm wondering if you considered (and how you dealt with) issues with consistency?

**Comment by liborw** [comment]
@gavanderhoorn: yes, the motoman side is [here](http://clopema.felk.cvut.cz/gitweb/?p=ros_industrial.git;a=tree;f=motoman/dx100/motoplus/motoros_lib;h=28bed42d0d9d3eee28a140097a51afb9a79fb7b5;hb=HEAD) io_interface.[cpp,h], and the message is [here](http://clopema.felk.cvut.cz/gitweb/?p=ros_industrial.git;a=tree;f=industrial_core/simple_message;h=4f34c2d9e9d6414fd11d2dd76e9e67aa0099ad22;hb=HEAD) io_data.[cpp,h]

**Comment by shaun-edwards** [comment]
@liborw, @gavanderhoorn, It seemed like this could be the start of a new feature discussion.  Was there any resolution on this.  This is definitely a feature of interest to a lot of users, but we (SwRI) doesn't have the resources to put to it.

**Comment by Samsagax** [comment]
I looked into @liborw code and I think I can implement something for an ABB robot. Maybe we should add a standard configuration file to map I/O units and addresses to human readable form.
Or perhaps we can add identifiers instead of I/O addresses, but that would depend greatly on each manufacturer's controller capabilities.

**Comment by shaun-edwards** [comment]
@Samsagax, The difficulty in this issue is standardizing IO addressing across robot platforms.  @gavanderhoorn is currently putting together a plan for this.  While we could implement specific solutions for different robots, it makes sense to take a generic approach so we can standardize at a driver level.  This will allow for code portability across vendor platforms.

**Comment by Samsagax** [comment]
> @gavanderhoorn is currently putting together a plan for this.

@gavanderhoorn, are you open for help?

**Comment by shaun-edwards** [comment]
@gavanderhoorn, ping...it looks like @Samsagax is willing to help out.

**Comment by gavanderhoorn** [comment]
@shaun-edwards: I'd seen it, just hadn't had time to respond yet.

@Samsagax: please contact me off-list (to avoid cluttering this issue).

**Comment by gavanderhoorn** [comment]
See also ros-industrial/rep#7.
